src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -1182,24 +1213,19 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                     bool b = (op_value.real() == 0.0 && op_value.imag() == 0.0);
                     value = ASR::down_cast<ASR::expr_t>(
                         ASR::make_LogicalConstant_t(al, x.base.base.loc, b, logical_type));
-                    operand_type = logical_type;
                 } else {
                     switch (op) {
                         case (ASR::unaryopType::UAdd): { result = op_value; break; }
                         case (ASR::unaryopType::USub): { result = -op_value; break; }
-                        default: {
-                            throw SemanticError("Bad operand type for unary " +
-                                ASRUtils::unop_to_str(op) + ": " + ASRUtils::type_to_str_python(operand_type),
-                                x.base.base.loc);
-                        }
+                        default: { LFORTRAN_ASSERT(false); } // should never happen
                     }
                     value = ASR::down_cast<ASR::expr_t>(
                         ASR::make_ComplexConstant_t(al, x.base.base.loc,
                         std::real(result), std::imag(result), operand_type));
                 }
             }
         }
-        tmp = ASR::make_UnaryOp_t(al, x.base.base.loc, op, operand, operand_type,
+        tmp = ASR::make_UnaryOp_t(al, x.base.base.loc, op, operand, result_type,
                               value);
```

Comment:
I think that this will take `not 5` as an Integer operand and return a Logical. That is not the way to do it.

We have to treat UnaryOp just like we treat BinOp. It accepts and returns exactly the same type, always. And we use `Cast` to cast things correctly. So `not 5` should be represented as `UnaryOp(Cast(5, IntToLogical))`. 

Also, in our ASR refactoring, we split UnaryOp into `UnaryMinus` (accepts integer, float, complex and returns the same type) and `LogicalNot` (accepts and returns logical) and `IntegerBitNot` (I think that is the "Invert" thing). So that is how it should be implemented.